core/arch/asm_defines.asm

```diff
@@ -903,6 +929,16 @@ ASSUME fs:_DATA @N@\
 # endif
 # define INC(reg) add reg, reg, POUND 1
 # define DEC(reg) sub reg, reg, POUND 1
+#elif defined(RISCV64)
+# define REG_SCRATCH0 REG_A0
+# define REG_SCRATCH1 REG_A1
+# define REG_SCRATCH2 REG_A2
+# define JUMP     j
+/* FIXME i#3544: RFC: There is no state register with compare flags. Emulate? */
+# define JUMP_NOT_EQUAL     bnez x5,
```

Comment:
Not really. Looking at the code again this macro is only used in the `thread_check_gprs_from_cache` test in `suite/tests/api/detach_state.c`:
```
check_gprs_from_cache_spin:
        CHECK_SIDELINE_EXIT
        JUMP_NOT_EQUAL check_gprs_from_cache_spin
        PUSHALL
```
The `CHECK_SIDELINE_EXIT` is defined in the same file.
Given that I see 2 solutions:
1. Remove `JUMP_NOT_EQUAL` from `asm_defines.asm` and define it straight in `suite/tests/api/detach_state.c`. This will remove the issue for repo-internal code.
2. Modify the `JUMP_NOT_EQUAL -> JUMP_NOT_EQUAL(reg_lhs, reg_rhs, label)`.

Both will be problematic if `asm_defines.asm` is an API file for extensions. If it is though, I think we're missing a `COMPARE(lhs,rhs,cmp)` macro.

I'll go with the first approach for now.

Reply:
This change looks good, though I'm trying to understand whether asm_defines is exported to users outside DR or not. If it is, then this change is compatibility breaking and we should preserve the existing JUMP_NOT_EQUAL. 

I see drmemory uses asm_defines directly: https://github.com/DynamoRIO/drmemory/blob/a887f796321b1d29b14f617d856a9f157f018c51/CMakeLists.txt#L1022.


Maybe the safest way is to define a new JUMP_NOT_EQUAL macro in asm_defines that takes the extra args for RISC-V and preserve the existing definitions for other archs.

Reply:
Done. RISC-V now has a separate `JUMP_NOT_EQUAL` and the code in `detach_state.c`.